### ISSUE TITLE
Convert accessor to pointer

### ISSUE THREAD

**Issue by szellmann** [amd/triSYCL#11]
Hi, I wonder how one may convert an accessor to a pointer. The [SYCL 1.2 specification](https://www.khronos.org/registry/sycl/specs/sycl-1.2.pdf) states under 2.7 Memory objects (p. 27), that accessors ...

> can be converted within a kernel into C++ pointer types, or the cl::sycl::global_ptr, cl::sycl::constant_ptr classes.

A quick grep for _global_ptr_ however reveals that this type is not present in the source code, and buffer/accessor have no ```operator T*``` for implicit conversion, or smth. similar. Is there a way to convert accessors to raw pointers with triSYCL?

For my purposes it would simplify things if one could obtain a pointer to buffer data on the host, similar to device pointers with CUDA. But that is probably not intended by the standard, right?

**Comment by keryell** [comment]
You are right it is not implemented yet. Since I was pushing inside the committee to have global<T*> instead of global_ptr<T>, I did not make it a top priority.
It should not be long to fix but I am too busy with SYCL 2.x... :-)

In the meantime on the device side you can always have something like:
auto p = &some_accessor[0][0];
But I am not sure it will solve your issue if you want to have it on the host side...
On some tests I use on the host for verification:
  for (auto const &e : c.get_access<cl::sycl::access::read>()) {
    // The difference between elements reconstructs the index value
    BOOST_CHECK(e == &e - &*c.get_access<cl::sycl::access::read>().begin());
  }
So you can get the address on an accessor element but not on a buffer.
In OpenCL there is no equivalent to the device pointer, except if you use SVM, which is OpenCL 2.x...

**Comment by szellmann** [comment]
Ah, thanks for the info.

Do you think that accessing the storage array through the begin() iterator is portable? Or is it something that might work by accident with certain implementations? I have to admit that I haven't worked all the way through the specification yet, but I'd rather believe they would have specified access to the device DDR memory in a more explicit way.

Excited about the upcoming SYCL 2.0 support, and even more excited about a first device implementation that I believe you mentioned elsewhere :)

**Comment by keryell** [comment]
Good question. Keep asking. :-) Accessing the storage array through the begin() iterator is not yet portable since it has been forgotten in the current specification but I think it will be accepted in the next update. Unfortunately for you, it is often difficult to learn only from the specifications, specially when they are still to be improved. This is why we need to do a lot of tutorials and examples... :-)

**Comment by andrewrichards** [comment]
In SYCL v1.2, you *can* convert an accessor or global_ptr to an actual pointer, *but* only on-device. The key thing is that the transfer from host-to-device *must* go through an accessor. But, on device, you can convert into a pointer. This is what the spec says, although whether that works on a specific implementation is another matter.

The issue is that when sharing data between host and device (or even between more than one device) there is no expectation that the pointers will have the same underlying value, or even be pointing to the same address space or have the same number of bits. This is what allows sharing of data across a heterogeneous system with different physical memories.

Once on device (i.e. within a parallel_for) you can use either C++ pointers, or the SYCL pointer classes.

However, what you can do with the pointers may be limited. For example, you cannot take a pointer on-device and store it in global memory to be re-used later. Because there is no guarantee that the pointer value will still be valid on a different kernel invocation or on a different device.

Some of these restrictions may disappear with OpenCL 2.x, however you do need to watch out that the shared-virtual-memory of OpenCL 2.0 has different capabilities on different operating systems and devices.

**Comment by szellmann** [comment]
Thank you for the advice. Actually, the reason why I'm interested in the pointer is that I have a struct that stores pointers to lists of vectors, colors etc. that I manage on the host. Now I have two code paths that use these pointers - one SSE code path - the pointers point to the heap, and a second code path, a CUDA kernel where memory is managed using thrust device vectors and the pointers point to GPU memory. Both code paths implement the same algorithm, but are however optimized for either the CPU or the GPU. I'm now porting to SYCL and am wondering if I can achieve smth. similar w/o to much code rewrite.

Sorry for hijacking the issue tracker - however I think it's best to keep the issue open because global_ptr (or global<T*>, for that matter) isn't implemented. However, is anyone of you aware of a place with active discussion on SYCL? I found the Khronos board to be rather dead.

**Comment by andrewrichards** [comment]
That sounds like something you can't do with SYCL 1.2, because it requires the pointers to be the same between host and device. That is an OpenCL 2.0 feature and not as widely supported.

You will need to use a buffer on the host. The advantage of buffers over pointers is that they manage scheduling dependencies and data movement for you as well as referencing the data. But it does mean changing your code so your data isn't just referenced by a pointer.

**Comment by keryell** [comment]
OK, now I understand better the bigger picture, I might answer the question. :-)

Actually this is something I talked about in a few talks in a slide with title "Poor-man SVM with C++11 + SYCL" you can look for with a search engine. For example in IWOCL 2015 tutorial http://codeplaysoftware.github.io/iwocl2015/ there are my slides and video. Slide 36 from http://codeplaysoftware.github.io/iwocl2015/presentations/2015-05-10-IWOCL_OpenCL_modern_C++-expose.pdf and at 28:15 in the video.
You need to add a little bit more code but it should not be that bad and globally more efficient than using SVM in some cases.
C++11 allocators can help a lot there.
Basically, to implement it, you just need to override the pointer operation when the preprocessor  symbol __SYCL_DEVICE_ONLY__ is defined to do the address translation on the fly.
Or in the more general case, you can have different implementations of your classes according to __SYCL_DEVICE_ONLY__.
Well actually you could do it systematically even when targeting the host (CPU) but you would end with a code to do a useless address translation by 0. At least it is useful for debugging the concept.

**Comment by szellmann** [comment]
Yes, this sounds just like the solution to my problem.

The problem that I'm facing is that buffers and accessors are two separate concepts, with the first being exclusive to the host (from an access point of view), and the second exclusive to the device. What I need is a single entity that behaves like a buffer on the host, that has random iterator semantics on the device (plus derefencing of course). Efficient random access on the host would be a plus but not mandatory.

What you're suggesting is: implement a custom allocator which manages a sycl buffer and provides a sycl accessor when calling allocator::address(). Am I right?

I don't quite get what you say in the presentation about the address mod a translation, but with the implementation you suggest, I would not need to care about different addresses (of course, host and device have different address spaces).

Nice talk. C++ is a completely different beast compared to the status quo from ten years ago, which many people from the HPC community haven't realized yet. I wasn't aware of that OpenCL 2.1 will add C++11 in kernel code, which is great.

**Comment by keryell** [comment]
The buffer is the abstract memory. Even on the host you have to access it though a host accessor (actually there are some conditions when you can access the memory otherwise but it is outside the scope of the discussion).

You are right, the buffer concept exists only on the host side, because only the host can manage memory for now. But the accessor provides you "random iterator semantics on the device (plus derefencing of course)". In my second comment of this discussion you can see the use of a host accessor iterator (well a little bit contrived...).

I was wondering what I could talk about during my future presentation at http://conf.researchr.org/track/PPoPP-2016/SYCL-2016-papers#program and you gave me the inspiration: transform the terse slide "Poor-man SVM with C++11 + SYCL" into a presentation. :-)

The difficulty is to hide everything in some nice high-level classes, which is not obvious to me yet.
But at the C level instead of C++, basically what has to be done:
- you create a buffer with some pointers inside pointing to some element inside;
- you get a host accessor and you keep somewhere the address of the first element of the accessor: it will be the host base address h_ba;
- on the device you have the device accessor and you start by measuring the address of the first element of the accessor and store it in a local variable, the device base address, d_ba;
- you overload all the pointer dereference *p to the accessor by *(p + d_ba - h_ba) and replace all the &e of an accessor element by &e - d_ba + h_ba so you can even update the pointers in the device accessor and it will work later on the host accessor;
- as an optimization, you store d_ba - h_ba on the device instead of only d_ba.

Now, how to hide this to the programmer in an elegant C++ way?
I am waiting for your suggestions. :-)

Of course having coarse grain SVM in next version of SYCL will alleviate all this issue but with the implication of having the same pointer size for both host and device, which may be overkill on small devices, where with this poor man version of SVM you may have 64 bit pointers on the host and say 16 bit pointers on the device.

**Comment by szellmann** [comment]
Thanks for helping me with this, your advice is much appreciated.

You're right, random-access is only part of what I need, and the accessor is sufficient for that.

I'll try to briefly state the problem I'm having (or think I'm having) with buffers and accessors. What I want to accomplish is to provide the user of my library with the following API:

```
typedef cuda_sched<...> scheduler;  // schedulers are classes that create threads
                                    // there's also a cpu_sched<>, and (soon) a sycl_sched :)
                                    // my user can choose the one that best suits her needs.

struct kernel                       // kernel is smth. that the user writes
{                                   // schedulers can execute kernels, each thread
                                    // basically calls operator()
    Result operator()(...)
    {
         // User code in here (some user defined algorithm that each thread computes)
         do_smth_with(data);
         or_with(other_data);
    }
    SomeType* data;                 // what data is used is completely up to the user
                                    // and the user can have as many pointers as she wants
};


//------------------------------------------------------
// some function scope

kernel kern;

// in this example, Cont would probably be an std::vector
// but it can generally be any "user-mode" data entity
// such as a raw pointer
Cont host_data

// This is the CUDA case, in the SSE/AVX case I'd just
// assign host_data.data() to kernel.data
thrust::device_vector d_data(host_data);
kern.data = thrust::raw_pointer_cast(d_data.data());

sched.frame(kern);                  // frame() "executes" the kernel (call operator())
                                    // frame() does all the magic behind the scene - e.g.
                                    // create a device kernel and a grid of threads
                                    // w/ cpu_sched (and newest CUDA), kernel could
                                    // also be a lambda function
```

A more thorough description of this can be found here: https://github.com/szellmann/visionaray/wiki/A-minimal-working-example (kernels are actually "ray tracing kernels", I left this out because it's irrelevant here).

Now, the user can basically just alter the typedef in the first line, handle data copies differently and target a completely different platform with the same code. As you can see, the user is responsible for copying data. What I **don't want** to place on the shoulders of the user (because I think it is not platform independent), is smth. like the following:

```
...
queue q;
q.submit([&](handler& cmdgroup)
{
    // derive an accessor from data
    kern.data = sycl_data.get_access()

    sched.frame(kern);
});
```

I don't want the user to care about command queues, because this only applies to sycl and is irrelevant in all other cases I can think of. What I'd thus rather do is create the command queue inside (my own class) sycl_sched's ```frame()``` function, and the user creating a SYCL buffer, and providing "access" to it on the host (outside the command queue). But then (if I get the SYCL thing completely right, which I'm not sure I do), I (as a lib developer) need to a priori know what data the user's kernel contains, which I don't. 

If I could obtain a device pointer (or smth. with similar semantics) from a buffer outside the command queue, that would help me much. I could also live with kernel.data not being a pointer but some C++ container template from my library.

Btw, I'd like to mention that I'm not completely committed to SYCL. I also did some experiments with C++ AMP, but this has other shortcomings that I dislike :) . I'm investigating this because I want to support AMD GPUs in addition to NVIDIA GPUs. Kernels are intended to be written in modern (C++11 or higher) C++.

**Comment by keryell** [comment]
Ouch, large answer! :-)
OK, perhaps your problem is simpler and you do not need address translation, just need a pointer to some data on the device.
The difficulty in OpenCL compared to CUDA is that in OpenCL they decided to be more abstract and have a cl_mem type while on CUDA you have a pointer, which is a priori invalid on the host but valid on the device.
This explains why
```
kern.data = thrust::raw_pointer_cast(d_data.data());
```
is trivial to implement: just a cast

In OpenCL you can do this by launching a kernel just to take the device address of the first element of the buffer and write it back in an auxiliary buffer so you can get it on the host for future use case.
But actually I think it is unspecified in OpenCL whether you will have always the same address for the buffer... So I wonder if your method can work in the general case...

In SYCL, actually there is no guarantee that every time a kernel is launched using a buffer that it will be under the hood the same OpenCL buffer and then it will be also the same address. But I cannot see why a sane implementation would have this spam of buffer creations... This is something we could clarify in the SYCL committee. But anyway, even the same OpenCL buffer may be remapped at different addresses on the device when used back and forth between host and device...

I will add an example in triSYCL test cases on this, since it seems an interesting point in some applications like yours, even if it is awful from a pure clean C++ point of view. :-)

So the cleaner way to achieve a sound behaviour is to use coarse grain shared virtual memory. There is no the potential execution overhead for each access of real shared memory given by fine grain shared virtual memory compared to plain buffers but at least you know that the addresses on the host and the device are the same. You just pay a bigger cost once when you transfer globally the data.

Otherwise, instead using low-level pointers like in CUDA, use something higher-level, basically closer to SYCL accessors, but that means to rethink a little bit your software architecture. Pointers are evil. :-) For example use always buffers and accessors, with accessors basically doing nothing on x86 and CUDA but just using SYCL accessors on SYCL platform.

In CUDA the default queue is the default stream which is hidden by default to the user. In SYCL we tried to remove a lot of magic, be cleaner and more explicit. The idea is to be thread-safe, for example. As you said you can simulate the same behaviour by keeping a SYCL queue in your SYCL scheduler.

Related to C++AMP, since it is older than SYCL and we took all (hopefully) the good ideas from it into SYCL and had a lot of other stuff, I do not really see the advantage you would have using it. You would have the same issue with array_view<> instead of accessor<> because there is no guarantee that the address is the same on host and device.

At the end, if you are lazy, you could just use the fine grain system or buffer shared virtual memory from OpenCL 2 that would land in next version of SYCL soon. Of course in triSYCL it works already by using only the host device, ie the CPU. ;-) My understanding is that AMD/MultiCoreWare hcc can provide you some fine grain memory too on the supported platforms, compared to plain old C++AMP.

Globally it would be interesting to benchmark the various solutions on your application. :-)
